File: Java/P007.java
// Project Euler
// Problem 7: The 10001st prime number

public class P007 {
    public static void main(String[] args) {
        int n_th = 10001;
        int range = n_th * 1000;    // assumming if the n_th prime is within the range of n_th * 1000

        while (!eratosthenes_sieve(n_th, range))
            range *= 1000;  // increase the range in case we don't get the n_th prime
    }

    public static boolean eratosthenes_sieve(int n_th, int range) {
        int[] array = new int[range];

        array[0] = 0;

        // filling the array
        for (int i = 1; i < range; i++)
            array[i] = i + 1;

        // the sieve
        for (int i = 1; i < range; i++) {
            if (array[i] != 0) {
                for (int j = 2*array[i]; j <= range; j += array[i])
                    array[j - 1] = 0;
            }
        }

        // checking
        int prime_th = 0;
        int index;
        for (index = 0; index < range; index++) {
            if (array[index] != 0) prime_th += 1;
            if (prime_th == n_th) {
                System.out.println(array[index]);
                return true;
            }
        }

        return false;
    }
}

File: Active/C++/p007.cpp
// Project Euler
// Problem 7: The 10001st prime nubmer

// I think it would be better if I use vector rather than array
// but at the time I don't understand about vector yet

#include <iostream>
using namespace std;

int Eratosthenes_Sieve(int limit, unsigned long int size)
{
    int Array[size];

    Array[0] = 0;   // it should be 1 but we assign it 0 in the first place
    for (int i = 1; i < size; ++i) Array[i] = i + 1; // filling the array

    // The sieve
    for (int i = 1; i < size; ++i) {
        if (Array[i] != 0) {
            for (int j = 2*Array[i]; j <= size; j += Array[i])
                Array[j - 1] = 0;
        }
    }

    // To check if n-th(limit-th) prime is on our limit. If not we multiply
    // the size of our array by 10. line 46 and 54
    int prime_th = 0;
    int index;
    for (index = 0; index < size; index++) {
        if (Array[index] != 0) prime_th += 1;
        if (prime_th == limit) break;
    }

    if (prime_th == limit) {
        cout << Array[index] << endl;
        return 1;
    }
    else return 0;  // there no n-th prime that we looking for
                    // in the range of size
}

int main()
{
    int limit = 10001;
    unsigned long int size = limit * 10;    // assuming the limit-th prime within the range of limit*10
    cout << "The 10001st prime number." << endl;

    while (!Eratosthenes_Sieve(limit, size)) {
        size *= 10;   // assuming the limit-th prime will be in the range of limit*10
    }

    cout << "Input another limit other than 10001.\n";
    cin >> limit;
    size = limit * 10;

    while (!Eratosthenes_Sieve(limit, size)) {
        size *= 10;   // assuming the limit-th prime will be in the range of limit*10
    }

    return 0;
}
